src/ncp/ncp_base_radio.cpp

```diff
@@ -446,22 +446,22 @@ otError NcpBase::HandlePropertySet_SPINEL_PROP_STREAM_RAW(uint8_t aHeader)
 
     SuccessOrExit(error = DecodeStreamRawTxRequest(*frame));
 
-    // Cache the transaction ID for async response
-    mCurTransmitTID = SPINEL_HEADER_GET_TID(aHeader);
-
     // Pass frame to the radio layer. Note, this fails if we
     // haven't enabled raw stream or are already transmitting.
-    error = otLinkRawTransmit(mInstance, &NcpBase::LinkRawTransmitDone);
+    SuccessOrExit(error = otLinkRawTransmit(mInstance, &NcpBase::LinkRawTransmitDone));
+
+    // Cache the transaction ID for async response
+    mCurTransmitTID = SPINEL_HEADER_GET_TID(aHeader);
 
 exit:
 
-    if (error == OT_ERROR_NONE)
+    if (error != OT_ERROR_NONE)
     {
-        // Don't do anything here yet. We will complete the transaction when we get a transmit done callback
-    }
-    else
-    {
-        error = WriteLastStatusFrame(aHeader, ThreadErrorToSpinelStatus(error));
+        // If we fail to report the error now, it will be reported later in HandleReceive() of ncp_base.cpp.
+        if (WriteLastStatusFrame(aHeader, ThreadErrorToSpinelStatus(error)) == OT_ERROR_NONE)
+        {
+            error = OT_ERROR_NONE;
+        }
```

Comment:
I think this may work  (keep the error in this case if we cannot append it here to be reported by next layer) but in original code we would get a `nobufs` error instead. This should only happen if we are out space in spinel buffer (cannot append the "last status" here). 

Though I must say we should be careful here. At some point we were considering to not allow the special "set handlers" to return any error (which is not the case in the code now). 

The question for me is if there is value to report the original error (probably "parse" or "invalid state") instead of "nobufs" (what we get from `WriteLastStatusFrame()`)? If not, maybe better to not change it. Thoughts?

Reply:
I think the function returns the error code of itself, not the intermediate status. The original error is more for the host side.

Reply:
Hi @abtink and @bukepo. I believe HandlePropertySet_SPINEL_PROP_STREAM_RAW should only return the error status of the transmit, not that of WriteLastStatusFrame which returns the error status of a write SPINEL message.  The error gets reported up to the host via a SPINEL message in one of two ways.  Either WriteLastStatusFrame succeeds, in which case the transmit error is reported up via SPINEL immediately, or it fails, in which case the error returned is prepared and enqueued into a SPINEL response message, to be sent to the host later.  Either way, what the host cares about is the error reported from the call to transmit.  Am I missing something?

Reply:
> I believe HandlePropertySet_SPINEL_PROP_STREAM_RAW should only return the error status of the transmit, not that of WriteLastStatusFrame

@mnp222 this is not the intention behind `HandlePropertySet_SPINEL_PROP_STREAM_RAW()` method.

This method is part of "special" set of spinel "property handlers" which are expected to prepare their own response:
https://github.com/openthread/openthread/blob/87284f22eb7a03865134de7b02b4425f93241868/src/ncp/ncp_base.cpp#L963-L1001

The change in this PR can technically work with current code, but as I mentioned we need to be careful, as it is not the intended model for these special property setters (the way their error is handled may change in future). So my recommendation is not to rely on this.

Reply:
Thanks for the comments, everyone. I did some analysis on the impact of different error code returns from HandlePropertySet_SPINEL_PROP_STREAM_RAW (). Reverted the changes to keep the original code flow since HandleTransmitDone() from host does not care about what error sent by rcp if it is not from LinkRawTransmitDone() callback. Return codes from HandlePropertySet_SPINEL_PROP_STREAM_RAW() would be ended become a OT_ERROR_PARSE on host.
Please review other changes in this PR.